Add support for import map and import prefix

jonathaningram opened the pull request:
*Issue #, if available:*

Refs #13.

*Description of changes:*

Provides the same support that `protoc-gen-go` supports for import mapping and import prefix.

I haven't written any docs on this...

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

spenczar commented:
As we discussed in #13, I think it's better to encourage people to use `option go_package` to indicate how things should be imported.

spenczar commented on protoc-gen-twirp/generator.go:
```diff
@@ -61,14 +64,43 @@ func newGenerator() *twirp {
+		case "import_prefix":
+			t.importPrefix = v
+		default:
+			if len(k) > 0 && k[0] == 'M' {
```
Ugh, this is a terrible API for users... but it _is_ what `github.com/golang/protobuf/protoc-gen-go` does. I suppose we should support it.

But can we also provide a less-insane format? I don't know exactly what, but here's an idea:
```
protoc --twirp_out="go_package@rpc/empty.proto=github.com/twitchtv/rpc/empty,go_package@something/else.proto=golang.org/x/net/notreal.proto:.
```

spenczar commented on protoc-gen-twirp/generator.go:
```diff
@@ -61,14 +64,43 @@ func newGenerator() *twirp {
+// commandLineParameters breaks the comma-separated list of key=value pairs
+// in the parameter (a member of the request protobuf) into a key/value map.
+// It then sets file name mappings defined by those entries.
+func (t *twirp) commandLineParameters(parameter string) {
```
Instead of making this a method on `*twirp`, could you make this a function at the top level that returns a struct that has the fields that come from the command line? I think that would make it easier to write unit tests and also make it easier to document and keep track of which parameters are set in the command line like this. 

I'm picturing the call site as something like this: 
```
params, err := parseParams(in.GetParameters())
if err != nil {
  gen.Fail("invalid parameters passed to --twirp_out: %v", err)
}
t.importMap = params.importMap
t.importPrefix = params.importPrefix
```

spenczar commented on protoc-gen-twirp/generator.go:
```diff
@@ -61,14 +64,43 @@ func newGenerator() *twirp {
+	param := make(map[string]string)
+	for _, p := range strings.Split(parameter, ",") {
+		if i := strings.Index(p, "="); i < 0 {
+			param[p] = ""
```
Eh, I think we can error on this. Let's only support `k=v`-style parameters.

jonathaningram commented:
@spenczar cheers, will address.

jonathaningram commented:
Quick update: I think I've addressed all feedback except the docs. I went with your suggestion for `go_package@`. A better option is not really apparent to me at the moment...

jonathaningram commented:
@spenczar thinking about the prefix to use for this, `go_package` is probably confusing as it conflicts with the option used in the proto file and also doesn't really say what it's doing.

What about `go_import_mapping`? So the option looks like:

```
go_import_mapping@rpc/empty.proto=github.com/user/rpc
```

1. It has a `go` prefix since this is for Go things.
2. The parameter prefix maps back to the Go variable name we actually use in the generator which is named `importMap`
3. In this way, the protoc-gen-go `M` actually becomes a shorthand flag despite missing the `@`.

jonathaningram commented:
@spenczar ok that's enough PR spam for you. Could you please review? I've added a doc file and also decided to change to `go_import_mapping` per the above. I think it makes much more sense.
